habitat_sim/simulator.py

```diff
@@ -73,18 +75,14 @@ def reconfigure(self, config: Configuration):
         if self.config == config:
             return
 
-        for agent in self.agents:
-            agent.detach()
-
         # NB: Configure backend last as this gives more time for python's GC
         # to delete any previous instances of the simulator
-        self._config_agents(config)
+        # TODO: can't do the above, sorry -- the Agent constructor needs access
+        # to self._sim.get_active_scene_graph()
```

Comment:
Doesn't seem to be a problem to me.  In any case, this is just eagerness on our end to make it easier for python GC, and there is no correctness or stability concern AFAIK, right @erikwijmans ?  So we can just remove the note above.

Reply:
There is a stability concern.  I have seen crashes due to trying to having two OpenGL contexts in the same process when rapidly creating and delete instances of the simulator.  Python's GC doesn't get around to cleaning up the old instance before we try to create a new instance.

Reply:
Do we *need* to recreate the GL context, tho? I think it would make sense to keep it around all the time and recreate only the scene and GL objects.

Reply:
Yeah, it is currently a super-hacky way to control GPU memory usage.  Whenever we swap a scene, we just kill the entire simulator backend as that dump all assets off the GPU :)

Reply:
Could we use a singleton for the OpenGL context?  That way the context will survive through destruction of the simulator?

Reply:
Singleton as in making the `WindowlessContext` independent from the `Simulator` (instead being protected there) and assigning it to a global `gl_context = WindowlessContext()`, for example? Yes that would work.

> it is currently a super-hacky way to control GPU memory usage

If you delete the whole `Simulator` instance every time and keep the GL context alive, it will cascade-delete all GL objects anyway -- and that's a perfectly valid way to clean up. Not "super hacky" at all :)

Is this a blocking issue for this PR (will it cause crashes in regular usage)? I have [platform.{glx,egl}.WindowlessApplication](https://doc.magnum.graphics/python/magnum.platform.egl.WindowlessApplication.html) exposed in Python, but currently without the additions you have in WindowlessContext -- would need to integrate those first.

Reply:
If we can't rapidly delete and recreate instances of the simulator, this will causes crashes in regular uses, habitat-api kills and recreates the simulator whenever the scene swaps, this is the hacky part :).

We could do the singleton in C++ via something like `WindowlessContext::Instance()` (would have to check to see how that works with threads and multi-processing).

There are also other ways to handle this, like forcing python GC to run, or having an explicit "close" method on the c++ side instead of letting the destructor handle it (as then we won't be reliant on the whims of python GC).

Let's definitely keep pushing this along and if it does cause crashing issues, we will figure it out!

Reply:
Having the GL context "singleton" / global on the Python side is not an option (`gl_context = WindowlessContext()` as I mentioned above)? That feels the most clean to me.

Reply:
Yeah, a python Singleton would work also 🙂

Reply:
Agreed.  Allowing a python singleton for the context is a good abstraction imo.  It would be good to make its explicit construction from the python side optional though so that we don't introduce additional boilerplate that is necessary to create a simulator instance.